src/solver/optimize.cpp

```diff
@@ -1,14 +1,296 @@
+            x0 = amms[j].buy_reserve_amount;
+        if (amms[j].sell_token_index == orders[i].buy_token_index)
+            y0 = amms[j].sell_reserve_amount;
+        if (amms[j].buy_token_index == orders[i].buy_token_index)
+            y0 = amms[j].buy_reserve_amount;
+
+        if ((x0 <= 0) || (y0 <= x_))
+            continue;
+
+        // at this point, we have found a "matching" AMM
+        // we must satisfy x1 * y1 >= x0 * y0, where (x1,y1) is the new pool balance and (x0, y0) is the original pool balance
+        // Since it is a fill-or-kill sell order, we have x1 = x0 + (1-fee)y
+        // and y1 = y0 - x_
+        // Thus, maximally exploiting the pool, we get that x1 = x0 * y0 / y1,
+        // which implies that y = ((y0 * x0 / y1) - x0)  / (1-fee);
+        float_y1 = static_cast<boost::multiprecision::mpf_float>(y0 - x_);
+        float_y = (( static_cast<boost::multiprecision::mpf_float>(y0 * x0) / float_y1) - static_cast<boost::multiprecision::mpf_float>(x0)) / (1 - amms[j].fee_amount);
+        y = static_cast<boost::multiprecision::mpz_int>(float_y);
+
+        if (y > y_)
+            continue;
+        
+        // at this point, the trade can happen and so we now check if objective improves, and if so, compute the prices
+        temp = (y_ - y) * tokens[orders[i].sell_token_index].external_price + orders[i].fee_amount * tokens[orders[i].fee_token_index].external_price - orders[i].cost_amount * tokens[orders[i].cost_token_index].external_price;
+
+        if (temp > best_order.objective_value) {
+            // found better solution
+            best_order.objective_value = temp;
+            best_order.executed_order_idx = i;
+            best_order.used_amm_idx = j;
+            best_order.executed_buy_amount = x_;
+            best_order.executed_sell_amount = y;
+            if (tokens[orders[i].buy_token_index].normalize_priority) {
+                best_order.buy_token_price = 1.0;
+                best_order.sell_token_price = (static_cast<boost::multiprecision::mpf_float>(x_)) / (static_cast<boost::multiprecision::mpf_float>(y));
+            }
+            else {
+                best_order.sell_token_price = 1.0;
+                best_order.buy_token_price = (static_cast<boost::multiprecision::mpf_float>(y)) / (static_cast<boost::multiprecision::mpf_float>(x_));
+            }
+            best_order.buy_token_price = best_order.buy_token_price * pow(10, tokens[orders[i].buy_token_index].decimals);
+            best_order.sell_token_price = best_order.sell_token_price * pow(10, tokens[orders[i].sell_token_index].decimals);
+        }
+    }
+    return;
+}
+
+
+
+void generate_output(std::vector<Token> &tokens, std::vector<Order> &orders, std::vector<CP_AMM> &amms, ExecutedOrder &best_order, nlohmann::json &output)
+{
+    bool found;
+    boost::multiprecision::mpz_int temp;
+
+    found = (best_order.objective_value > 0) ? true : false;
+
+    output["amms"];
```

Comment:
Interesting, I suppose this is here so that the "amms" key is created in the output object, right?

Reply:
Yes, that's correct. Not sure if this is the "correct" way to do so, though.

Reply:
I think that works! The way I usually do is something in the direction of 
```c++
output.push("amms"); // not sure if "push" is the name of the function we want
```